File: 2024/day18.cpp
// https://adventofcode.com/2024/day/18

#include "../common/common.h"

#include <array>
#include <optional>
#include <print>
#include <ranges>
#include <span>
#include <string>
#include <string_view>
#include <type_traits>
#include <vector>

using memspace_t = aoc::char_grid<>;
constexpr const char corrupted = '#';
constexpr const char empty = '.';

template <int num_fallen>
constexpr memspace_t bytes_fall(memspace_t memspace,
                                std::span<const point> falling_bytes) {
  AOC_ASSERT(num_fallen <= falling_bytes.size(), "Too many bytes requested");
  for (const auto byte_pos : falling_bytes.subspan(0, num_fallen)) {
    memspace.modify(corrupted, byte_pos.y, byte_pos.x);
  }
  return memspace;
}

constexpr std::optional<int> shortest_path(const memspace_t& memspace) {
  const auto start_pos = point{0, 0};
  const auto end_pos =
      point(memspace.num_rows() - 1, memspace.num_columns() - 1);

  const auto distances = aoc::shortest_distances_dijkstra(
      start_pos, end_pos, [&](const point current) {
        return memspace.basic_neighbor_positions(current) |
               stdv::filter([&](point neighbor) {
                 return memspace.at(neighbor.y, neighbor.x) != corrupted;
               }) |
               aoc::dijkstra_uniform_neighbors_view();
      });

  auto it = distances.find(end_pos);
  if (it == std::end(distances)) {
    return {};
  }
  return it->second;
}

constexpr memspace_t bytes_fall_dyn(memspace_t memspace,
                                    std::span<const point> falling_bytes,
                                    int count) {
  for (const auto byte_pos : falling_bytes.subspan(0, count)) {
    memspace.modify(corrupted, byte_pos.y, byte_pos.x);
  }
  return memspace;
}

template <int num_fallen>
constexpr point first_problematic_byte(const memspace_t& memspace,
                                       std::span<const point> falling_bytes) {
  // Adding corrupted cells can only ever block a path, never reopen one,
  // so whether a path exists is monotonic in the number of fallen bytes
  // That lets us binary search for the first blocking byte
  // instead of testing every byte one at a time
  const auto has_path_after = [&](int count) {
    return shortest_path(bytes_fall_dyn(memspace, falling_bytes, count))
        .has_value();
  };

  // We know from part 1 that num_fallen bytes don't cause a problem yet
  int low = num_fallen;
  int high = static_cast<int>(falling_bytes.size());
  if (has_path_after(high)) {
    return {};
  }
  while (low + 1 < high) {
    int mid = low + (high - low) / 2;
    if (has_path_after(mid)) {
      low = mid;
    } else {
      high = mid;
    }
  }

  return falling_bytes[high - 1];
}

template <point grid_size, int num_fallen, bool check_closed>
auto solve_case(const std::string& filename) {
  std::vector<point> falling_bytes;

  for (std::string_view line : aoc::views::read_lines(filename)) {
    falling_bytes.push_back(aoc::split<point>(line, ','));
  }

  memspace_t memspace{
      stdv::repeat(empty, grid_size.y * grid_size.x) | aoc::collect_string(),
      grid_size.y, grid_size.x};

  auto result = std::conditional_t<check_closed, point, int>{};
  if constexpr (!check_closed) {
    result = *shortest_path(bytes_fall<num_fallen>(memspace, falling_bytes));
  } else {
    result = first_problematic_byte<num_fallen>(memspace, falling_bytes);
  }
  return result;
}

int main() {
  std::println("Part 1");
  AOC_EXPECT_RESULT(22, (solve_case<{7, 7}, 12, false>("day18.example")));
  AOC_EXPECT_RESULT(234, (solve_case<{71, 71}, 1024, false>("day18.input")));
  std::println("Part 2");
  AOC_EXPECT_RESULT((point{6, 1}),
                    (solve_case<{7, 7}, 12, true>("day18.example")));
  AOC_EXPECT_RESULT((point{58, 19}),
                    (solve_case<{71, 71}, 1024, true>("day18.input")));
  AOC_RETURN_CHECK_RESULT();
}

File: rust/2024/src/day18.rs
use aoc::dijkstra::DijkstraNeighborView;
use aoc::grid::{Grid, Upos};

const CORRUPTED: char = '#';
const EMPTY: char = '.';

fn parse(filename: &str) -> Vec<Upos> {
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| Upos::parse(line, ",").unwrap())
        .collect()
}

fn bytes_fall(mut memspace: Grid<char>, falling_bytes: &[Upos]) -> Grid<char> {
    for &byte_pos in falling_bytes {
        memspace.modify(CORRUPTED, byte_pos.y, byte_pos.x);
    }
    memspace
}

fn shortest_path(memspace: &Grid<char>) -> Option<u32> {
    let start_pos = Upos::new(0, 0);
    let end_pos = Upos::new(memspace.num_columns - 1, memspace.num_rows - 1);

    let distances = aoc::dijkstra::shortest_distances(
        &start_pos,
        |current: &Upos| *current == end_pos,
        |current: &Upos| {
            memspace
                .basic_neighbor_positions(*current)
                .into_iter()
                .filter(|neighbor| *memspace.get(neighbor.y, neighbor.x) != CORRUPTED)
                .dijkstra_uniform_neighbors()
        },
    );

    distances.get(&end_pos).copied()
}

fn solve_case1<const NUM_ROWS: usize, const NUM_COLUMNS: usize, const NUM_FALLEN: usize>(
    falling_bytes: &[Upos],
) -> u32 {
    let memspace = Grid::new(EMPTY, NUM_ROWS, NUM_COLUMNS);
    shortest_path(&bytes_fall(memspace, &falling_bytes[..NUM_FALLEN])).unwrap()
}

fn first_problematic_byte(
    memspace: &Grid<char>,
    falling_bytes: &[Upos],
    num_fallen: usize,
) -> Option<Upos> {
    // Adding corrupted cells can only ever block a path, never reopen one,
    // so whether a path exists is monotonic in the number of fallen bytes
    // That lets us binary search for the first blocking byte
    // instead of testing every byte one at a time
    let has_path_after = |count: usize| {
        shortest_path(&bytes_fall(memspace.clone(), &falling_bytes[..count])).is_some()
    };

    // We know from part 1 that num_fallen bytes don't cause a problem yet
    let mut low = num_fallen;
    let mut high = falling_bytes.len();
    if has_path_after(high) {
        return None;
    }
    while low + 1 < high {
        let mid = low + (high - low) / 2;
        if has_path_after(mid) {
            low = mid;
        } else {
            high = mid;
        }
    }

    Some(falling_bytes[high - 1])
}

fn solve_case2<const NUM_ROWS: usize, const NUM_COLUMNS: usize, const NUM_FALLEN: usize>(
    falling_bytes: &[Upos],
) -> Upos {
    let memspace = Grid::new(EMPTY, NUM_ROWS, NUM_COLUMNS);
    first_problematic_byte(&memspace, falling_bytes, NUM_FALLEN).unwrap()
}

fn main() {
    println!("Part 1");
    let example = parse("day18.example");
    aoc::expect_result!(22, solve_case1::<7, 7, 12>(&example));
    let input = parse("day18.input");
    aoc::expect_result!(234, solve_case1::<71, 71, 1024>(&input));

    println!("Part 2");
    aoc::expect_result!(Upos::new(6, 1), solve_case2::<7, 7, 12>(&example));
    aoc::expect_result!(Upos::new(58, 19), solve_case2::<71, 71, 1024>(&input));
}
